### PR TITLE
Even more BN hacks

### PR DESCRIPTION
**Pull request by alancucki**

### CODE REVIEW

File: data.py [hunk 1]
```diff
@@ -245,8 +262,8 @@ def sample_batch(self, bsz, sample_sentence=SAMPLE_SENTENCE):
         yield (batch_tensor, batch_tensor) #(source, target)
 
 
-class UTF8WordStarFile(object):
-    def __init__(self, path, cuda, p=0.5, max_w_len=1000, **kwargs):
+class UTF8WordStarFile(UTF8File):
+    def __init__(self, path, cuda, p=0.2, max_w_len=1000, **kwargs):
```
Comment (smalik169) [inline data.py]: chyba miało zostać p=0.5

File: models.py [hunk 2]
```diff
@@ -411,6 +458,58 @@ def eval_on(self, batch_iterator, switch_to_evalmode=True):
...
+            #
+            # NOTE If not halving/normalizing, then zero_var=False
+            reset_batchnorms(self, momentum=0.5, zero_var=True)
+
+            # Note that evaluation= controls shuffling
+            for i, (src, tgt) in enumerate(dataset.iter_epoch(bsz, evaluation=True, len_=L)):
+                src = Variable(src, volatile=True)
+                tgt = Variable(tgt, volatile=True)
+                self._encode_decode(src, tgt)
+                apply_to_batchnorm(self, double_running)
```
Comment (smalik169) [inline models.py:489]: Tego nie rozumiem, dlaczego podwajać te statystyki?

File: models.py [hunk 3]
```diff
@@ -201,11 +226,13 @@ def forward(self, x, r, embed=True):
 class ByteCNNDecoder(nn.Module):
     def __init__(self, n, emsize, vocab_size, normalization,
                  use_linear_layers=True, compress_channels=None,
-                 output_embeddings_init=None):
+                 output_embeddings_init=None,
+                 use_external_batch_norm=False, external_batch_max_r=None):
```
Comment (smalik169) [inline models.py:230]: W dekoderze to chyba nie ma sensu, chyba, że chcesz zrobić wersję "kwadratową".

File: models.py [hunk 4]
```diff
@@ -247,8 +274,17 @@ def __init__(self, n, emsize, vocab_size, normalization,
...
-                *(residual_list(conv_proto, n//2, normalization,
+                *(residual_list(conv_proto, n//2,
+                                normalization=normalization,
                                 last_relu=(output_embeddings_init is not None))))
+        if use_external_batch_norm:
+            assert external_batch_max_r is not None, \
+                    'to use external batchnorm in reccurent part' + \
+                    ' you need to specify maximum number of recurrent steps'
+            self.rec_batchnorm_list = nn.ModuleList([
+                nn.BatchNorm1d(emsize, momentum=0.1) for _ in range(2 * (n // 2) * external_batch_max_r)])
```
Comment (smalik169) [inline models.py:286]: W obecnej wersji to nie bierze pod uwagę np. tego, że do pierwszego batchnorma wchodzą wszystkie dane (niezależnie od liczby obrotów rekurencji w enkoderze).

File: models.py (hunk 2 above)
Comment (alancucki) [reply models.py:489]: Chcę wziąć zwykłą średnią dla (mean,var) w BN. Średnia krocząca z momentum=0.5 i podwajaniem upraszcza się do zwykłej średniej.

To można zrobić czyściej, owijają warstwę BatchNorm w nową klasę.

File: models.py (hunk 3 above)
Comment (alancucki) [reply models.py:230]: W rzeczy samej, wersja kwadratowa.

File: models.py (hunk 4 above)
Comment (alancucki) [reply models.py:286]: To nie szkodzi; agregujemy statystyki dla każdej długości L osobno, czyli ze stałą ilością obrotów rekurencji.

Comment (smalik169) [reply models.py:286]: Czyli, jeżeli dobrze rozumiem, zmieniamy trochę semantykę `external batch norm`? Teraz używamy tego (tylko?) w kontekście przeliczania statystyk przy ewaluacji?

Comment (alancucki) [reply models.py:286]: Dokładnie tak. Podczas treningu i tak nie ma żadnej różnicy, czy masz jeden BN, czy kwadratowo wiele.
